File: cortex-py/cortex/query.py
"""Query filters for time-travel queries."""

from dataclasses import dataclass
from enum import Enum
from typing import Optional


class QueryOrder(str, Enum):
    ASC = "asc"
    DESC = "desc"


@dataclass
class QueryFilter:
    """Filter for querying decisions.

    Example:
        filter = QueryFilter.for_agent("my-agent").time_range("10:00", "12:00")
        decisions = db.query(filter)
    """

    agent_id: Optional[str] = None
    min_confidence: Optional[float] = None
    max_confidence: Optional[float] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    decision_type: Optional[str] = None
    limit: Optional[int] = None
    order: Optional[QueryOrder] = None

    @classmethod
    def for_agent(cls, agent_id: str) -> "QueryFilter":
        """Create a filter for a specific agent."""
        return cls(agent_id=agent_id)

    def time_range(self, start: str, end: str) -> "QueryFilter":
        """Set time range."""
        self.start_time = start
        self.end_time = end
        return self

    def confidence_range(self, min_val: float, max_val: float) -> "QueryFilter":
        """Set confidence range."""
        self.min_confidence = min_val
        self.max_confidence = max_val
        return self

    @classmethod
    def low_confidence(cls) -> "QueryFilter":
        """Filter for low confidence decisions (< 0.5)."""
        return cls(max_confidence=0.5)

    def limit_results(self, n: int) -> "QueryFilter":
        """Limit results."""
        self.limit = n
        return self

    def to_sql_where(self) -> tuple[str, list]:
        """Convert to SQL WHERE clause and params."""
        conditions = ["1=1"]
        params = []

        if self.agent_id:
            conditions.append("agent_id = ?")
            params.append(self.agent_id)

        if self.start_time:
            conditions.append("timestamp >= ?")
            params.append(self.start_time)

        if self.end_time:
            conditions.append("timestamp <= ?")
            params.append(self.end_time)

        if self.min_confidence is not None:
            conditions.append("confidence >= ?")
            params.append(self.min_confidence)

        if self.max_confidence is not None:
            conditions.append("confidence <= ?")
            params.append(self.max_confidence)

        if self.decision_type:
            conditions.append("decision = ?")
            params.append(self.decision_type)

        return " AND ".join(conditions), params

File: cortex-core/src/query.rs
//! Query filters for time-travel queries

use serde::{Deserialize, Serialize};

/// Filter for querying decisions
#[derive(Debug, Clone, Default)]
pub struct QueryFilter {
    /// Filter by agent ID
    pub agent_id: Option<String>,

    /// Minimum confidence threshold
    pub min_confidence: Option<f32>,

    /// Maximum confidence threshold
    pub max_confidence: Option<f32>,

    /// Start timestamp (ISO 8601)
    pub start_time: Option<String>,

    /// End timestamp (ISO 8601)
    pub end_time: Option<String>,

    /// Filter by decision type
    pub decision_type: Option<String>,

    /// Limit results
    pub limit: Option<usize>,

    /// Order by (asc/desc)
    pub order: Option<QueryOrder>,
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub enum QueryOrder {
    #[default]
    Desc,
    Asc,
}

impl QueryFilter {
    /// Create a new filter for a specific agent
    pub fn for_agent(agent_id: impl Into<String>) -> Self {
        Self {
            agent_id: Some(agent_id.into()),
            ..Default::default()
        }
    }

    /// Set time range
    pub fn time_range(mut self, start: impl Into<String>, end: impl Into<String>) -> Self {
        self.start_time = Some(start.into());
        self.end_time = Some(end.into());
        self
    }

    /// Set confidence range
    pub fn confidence_range(mut self, min: f32, max: f32) -> Self {
        self.min_confidence = Some(min);
        self.max_confidence = Some(max);
        self
    }

    /// Low confidence decisions (below 0.5)
    pub fn low_confidence() -> Self {
        Self {
            min_confidence: None,
            max_confidence: Some(0.5),
            ..Default::default()
        }
    }

    /// Set limit
    pub fn limit(mut self, n: usize) -> Self {
        self.limit = Some(n);
        self
    }
}
